File: server/sonar-web/src/main/js/components/controls/__tests__/Favorite-test.js
/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import mockery from 'mockery';
import chai, { expect } from 'chai';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import React from 'react';

chai.use(sinonChai);

function click (element) {
  return element.simulate('click', {
    target: { blur () {} },
    preventDefault () {}
  });
}

describe('Components :: Controls :: Favorite', () => {
  let addFavoriteStub;
  let removeFavoriteStub;
  let Favorite;
  let FavoriteInner;

  before(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    addFavoriteStub = sinon.stub().returns(Promise.resolve());
    removeFavoriteStub = sinon.stub().returns(Promise.resolve());

    mockery.registerMock('../../api/favorites', {
      addFavorite: addFavoriteStub,
      removeFavorite: removeFavoriteStub
    });

    Favorite = require('../Favorite').default;
    FavoriteInner = require('../Favorite').FavoriteInner;
  });

  after(() => {
    mockery.disable();
  });

  beforeEach(() => {
    addFavoriteStub.reset();
    removeFavoriteStub.reset();
  });

  it('should render FavoriteInner', () => {
    const favorite = shallow(
        <Favorite favorite={true} component="key"/>
    );
    const inner = favorite.find(FavoriteInner);
    expect(inner).to.have.length(1);
    expect(inner.prop('favorite')).to.equal(true);
    expect(inner.prop('onToggle')).to.be.defined;
  });

  it('should remove favorite', () => {
    const favorite = mount(
        <Favorite favorite={true} component="key"/>
    );
    click(favorite.find('a'));
    expect(removeFavoriteStub).to.have.been.calledWith('key');
  });

  it('should add favorite', () => {
    const favorite = mount(
        <Favorite favorite={false} component="key"/>
    );
    click(favorite.find('a'));
    expect(addFavoriteStub).to.have.been.calledWith('key');
  });

  it('should render favorite', () => {
    const favorite = shallow(
        <FavoriteInner favorite={true} onToggle={() => true}/>
    );
    expect(favorite.is('.icon-star-favorite')).to.equal(true);
  });

  it('should render not favorite', () => {
    const favorite = shallow(
        <FavoriteInner favorite={false} onToggle={() => true}/>
    );
    expect(favorite.is('.icon-star-favorite')).to.equal(false);
  });

  it('should call onToggle', () => {
    const onToggle = sinon.spy();
    const favorite = shallow(
        <FavoriteInner favorite={true} onToggle={onToggle}/>
    );
    click(favorite);
    expect(onToggle).to.have.been.called;
  });
});

File: tests/src/test/java/org/sonarqube/tests/analysis/FavoriteTest.java
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarqube.tests.analysis;

import com.sonar.orchestrator.Orchestrator;
import com.sonar.orchestrator.build.SonarScanner;
import org.sonarqube.tests.Category3Suite;
import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.ClassRule;
import org.junit.Test;
import org.sonarqube.ws.Favorites;
import org.sonarqube.ws.Favorites.Favorite;
import org.sonarqube.ws.Permissions;
import org.sonarqube.ws.client.WsClient;
import org.sonarqube.ws.client.favorites.SearchRequest;
import org.sonarqube.ws.client.permissions.AddProjectCreatorToTemplateRequest;
import org.sonarqube.ws.client.permissions.RemoveProjectCreatorFromTemplateRequest;
import org.sonarqube.ws.client.permissions.SearchTemplatesRequest;

import static com.sonar.orchestrator.container.Server.ADMIN_LOGIN;
import static com.sonar.orchestrator.container.Server.ADMIN_PASSWORD;
import static org.assertj.core.api.Assertions.assertThat;
import static util.ItUtils.newAdminWsClient;
import static util.ItUtils.projectDir;

public class FavoriteTest {
  @ClassRule
  public static Orchestrator orchestrator = Category3Suite.ORCHESTRATOR;
  private static WsClient adminWsClient;

  private static final String PROJECT_KEY = "sample";

  @Before
  public void setUp() {
    orchestrator.resetData();
  }

  @After
  public void tearDown() {
    removeProjectCreatorPermission();
  }

  @BeforeClass
  public static void classSetUp() {
    adminWsClient = newAdminWsClient(orchestrator);
  }

  @Test
  public void project_as_favorite_when_authenticated_and_first_analysis_and_a_project_creator_permission() {
    SonarScanner sampleProject = createScannerWithUserCredentials();
    addProjectCreatorPermission();

    orchestrator.executeBuild(sampleProject);

    Favorites.SearchResponse response = adminWsClient.favorites().search(new SearchRequest());
    assertThat(response.getFavoritesList()).extracting(Favorite::getKey).contains(PROJECT_KEY);
  }

  @Test
  public void no_project_as_favorite_when_no_project_creator_permission() {
    SonarScanner sampleProject = createScannerWithUserCredentials();

    orchestrator.executeBuild(sampleProject);

    Favorites.SearchResponse response = adminWsClient.favorites().search(new SearchRequest());
    assertThat(response.getFavoritesList()).extracting(Favorite::getKey).doesNotContain(PROJECT_KEY);
  }

  @Test
  public void no_project_as_favorite_when_second_analysis() {
    SonarScanner sampleProject = SonarScanner.create(projectDir("shared/xoo-sample"));
    orchestrator.executeBuild(sampleProject);
    sampleProject = createScannerWithUserCredentials();
    addProjectCreatorPermission();

    orchestrator.executeBuild(sampleProject);

    Favorites.SearchResponse response = adminWsClient.favorites().search(new SearchRequest());
    assertThat(response.getFavoritesList()).extracting(Favorite::getKey).doesNotContain(PROJECT_KEY);
  }

  private static SonarScanner createScannerWithUserCredentials() {
    return SonarScanner.create(projectDir("shared/xoo-sample"))
      .setProperty("sonar.login", ADMIN_LOGIN)
      .setProperty("sonar.password", ADMIN_PASSWORD);
  }

  private void addProjectCreatorPermission() {
    Permissions.SearchTemplatesWsResponse permissionTemplates = adminWsClient.permissions().searchTemplates(new SearchTemplatesRequest());
    assertThat(permissionTemplates.getDefaultTemplatesCount()).isEqualTo(1);
    adminWsClient.permissions().addProjectCreatorToTemplate(new AddProjectCreatorToTemplateRequest()
      .setTemplateId(permissionTemplates.getDefaultTemplates(0).getTemplateId())
      .setPermission("admin"));
  }

  private void removeProjectCreatorPermission() {
    Permissions.SearchTemplatesWsResponse permissionTemplates = adminWsClient.permissions().searchTemplates(new SearchTemplatesRequest());
    assertThat(permissionTemplates.getDefaultTemplatesCount()).isEqualTo(1);
    adminWsClient.permissions().removeProjectCreatorFromTemplate(new RemoveProjectCreatorFromTemplateRequest()
      .setTemplateId(permissionTemplates.getDefaultTemplates(0).getTemplateId())
      .setPermission("admin"));
  }
}
